Survey/Survey/Sources/Supports/Helper/KeyChain/KeychainHelper.swift

```
@@ -0,0 +1,118 @@
+            NSLog(error.localizedDescription)
+        }
+    }
+
+    private func save(data: Data) throws {
+        let query: [String: AnyObject] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrService as String: service as AnyObject,
+            kSecAttrAccount as String: account as AnyObject,
+            kSecValueData as String: data as AnyObject
+        ]
+        let status = SecItemAdd(query as CFDictionary, nil)
+        if status == errSecDuplicateItem {
+            throw KeyChainError.duplicateEntry
+        }
+        if status != errSecSuccess {
+            throw KeyChainError.unknown(status)
+        }
+    }
+
+    private func read() -> Data? {
+        let query: [String: AnyObject] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrService as String: service as AnyObject,
+            kSecAttrAccount as String: account as AnyObject,
+            kSecReturnData as String: kCFBooleanTrue,
+            kSecMatchLimit as String: kSecMatchLimitOne
+        ]
+        var result: AnyObject?
+        _ = SecItemCopyMatching(query as CFDictionary, &result)
+        return result as? Data
+    }
+
+    private func update(newData: Data) {
+        let query: [String: AnyObject] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrAccount as String: account as AnyObject
+        ]
+        let newAttributes: [String: AnyObject] = [
+            kSecValueData as String: newData as AnyObject
+        ]
+        let status = SecItemUpdate(query as CFDictionary, newAttributes as CFDictionary)
+        if status != errSecSuccess {
+            NSLog("cant update keychain")
+        } else {
+            NSLog("key chain data updated")
+        }
+    }
+
+    func delete() {
+        let query: [String: AnyObject] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrService as String: service as AnyObject,
```

- While some functions like `save(item:)`, and `update(item:)` are good, the others like `read()`, and `delete()` seem too short and do not convey enough their meaning and usage. Please consider more descriptive naming 🙏 
- There're 4 main operations: save, read, update and delete. These actions execute by processing the OS resources (keychain) and they can be failed. Consider making a clear result for the callers of these functions so they know the operation is executed successfully or failed, instead of making it silently. Also consider using the same, consistent API (i.e. returning type) for them. Two solutions I come up with are either return `Bool` or `throws`.

---

I suggest this [library](https://github.com/kishikawakatsumi/KeychainAccess) which is used in WORPT project

---

@ducbm051291 Agree 👍
I actually used that [KeychainAccess](https://github.com/kishikawakatsumi/KeychainAccess) in my code challenge while interviewing with Nimble 😄 And I'm going to use it in my current IC too!

---

done in this [commit](https://github.com/Shayokh144/ic-survey-swiftui-combine/pull/44/commits/f8b1f24fbec87c8dd42a17c31afcdba49baa20f7)